### PR TITLE
New client-side retry design

### LINKED ISSUE #51: Redesign client-side authentication flow + retry logic
## Context
Authentication on the client side just means sending an auth token with the auth wire format over the stream before anything else. This should happen at the socket-level because it should be transport-agnostic. However, because we currently handle TCP reconnects at the stream-level with stubborn IO abstractions, the stream also needs to run through the auth flow every time it reconnects (and thus needs to be aware of the necessary authentication).

This is not ideal, and for example doesn't fit well with QUIC streams (that automatically handle reconnects and thus only need to authenticate once). That's why QUIC doesn't support authentication yet.

## Proposal
Get rid of stubborn IO abstractions on such a low-level and introduce them at a higher level (in the socket). The transport abstraction is only used to get IO objects (normal `AsyncRead + AsyncWrite` streams). If one of these fails, the client socket will have some sort of `ConnectionManager` to retry the connection and run through the auth flow again. That way we don't leak this authentication abstraction into the lower levels.

### LINKED ISSUE #52: Support authentication for QUIC transport
## Context
We shipped QUIC support in #48 without authentication, because we need to redesign that system first (see #51). After #51 is implemented, support & test authentication.

### PR DESCRIPTION
**Pull request by mempirate**
Implements connection retries at the socket level, making it easier to do authentication & re-authentication.

This will however break retries for `ReqSocket`, which should be dealt with in a follow-up PR. The main change would be to move connection establishment inside of the driver instead of the socket handle and follow the same pattern as in `SubSocket`.

Closes #51
Closes #52

### CODE REVIEW

**Comment by mempirate** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #59** <a href="https://app.graphite.dev/github/pr/chainbound/msg-rs/59?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/59?utm_source=stack-comment).

File: msg-socket/src/req/socket.rs [hunk 1]
```diff
@@ -61,18 +64,39 @@ where
...
         let (to_driver, from_socket) = mpsc::channel(DEFAULT_BUFFER_SIZE);
 
         tracing::info!("Connecting to {}", endpoint);
 
-        let stream = self
-            .transport
-            .connect(endpoint)
-            .await
-            .map_err(|e| ReqError::Transport(Box::new(e)))?;
+        let mut backoff = ExponentialBackoff::new(Duration::from_millis(20), 256);
```
Comment (merklefruit) [inline msg-socket/src/req/socket.rs:74]: I suggest we cap the max retries to around 15?
that should be already enough where the last try will have 10min of backoff

Comment (mempirate) [reply msg-socket/src/req/socket.rs:74]: Yeah that's fair, 256 retries is infinity

**Review by merklefruit — APPROVED** [review]
Beautiful! One Q: do we lose the ability to blocking_connect or would it be possible to add a method to simply wait until the connection task is done (maybe via a reply channel)?

EDIT: for pub/sub I mean

File: msg-socket/src/req/socket.rs [hunk 2]
```diff
@@ -61,18 +64,39 @@ where
         response_rx.await.map_err(|_| ReqError::SocketClosed)?
     }
 
-    /// Connects to the target with the default options.
+    /// Connects to the target with the default options. WARN: this will block until the connection can be established.
```
Comment (merklefruit) [inline msg-socket/src/req/socket.rs:67]: nit: it's async so it won't block the thread, but rather awaits

Comment (mempirate) [reply msg-socket/src/req/socket.rs:67]: Resolved in upstack PR!

**Comment by mempirate** [comment]
### Merge activity

* **Jan 19, 9:46 AM**: @mempirate started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/chainbound/msg-rs/59).

**Pull request merged**